core/src/test/java/org/apache/iceberg/TestSchemaUpdate.java

```diff
@@ -2204,8 +2222,7 @@ public void testDeleteContainingNestedIdentifierFieldColumnsFails() {
                 new SchemaUpdate(newSchema, SCHEMA_LAST_COLUMN_ID + 2).deleteColumn("out").apply())
         .isInstanceOf(IllegalArgumentException.class)
         .hasMessage(
-            "Cannot delete field 24: out: required struct<25: nested: required string> "
-                + "as it will delete nested identifier field 25: nested: required string");
+            "Cannot delete field 26: out: required struct<27: nested: required string> as it will delete nested identifier field 27: nested: required string");
   }
 
```

Comment:
You introduce new functionally which should be covered by new test cases. At least this can be good cases to start :

- v2 tables rejecting date→timestamp promotion
- v3 tables rejecting date→timestamptz promotion
- Partitioned date fields rejecting promotion